typescript/sdk/src/providers/MultiProvider.ts

```
@@ -1,139 +1,272 @@
+   * @throws if chain's metadata has not been set
+   */
+  getDomainId(chainNameOrId: ChainName | number): number {
+    const metadata = this.getChainMetadata(chainNameOrId);
+    return metadata.domainId ?? metadata.id;
+  }
+
+  /**
+   * Get the ids for all chains known to this MultiProvider
+   */
+  getDomainIds(): number[] {
+    return this.getChainNames().map(this.getDomainId);
+  }
+
+  /**
+   * Get the chain name for a given domain id
+   */
+  domainIdToChainName(domainId: number): string {
+    const metadataList = Object.values(this.metadata);
+    const metadata =
+      metadataList.find((c) => c.domainId === domainId) ||
+      metadataList.find((c) => c.id === domainId);
+    if (!metadata) throw new Error(`No chain found for domain id ${domainId}`);
+    return metadata.name;
+  }
+
+  /**
+   * Get the chain id for a given domain id
+   */
+  domainIdToChainId(domainId: number): number {
+    const chainName = this.domainIdToChainName(domainId);
+    return this.getChainId(chainName);
+  }
+
+  /**
+   * Returns name for given domain id or name
+   */
+  resolveDomainOrName(domainIdOrName: ChainName | number): ChainName {
+    if (typeof domainIdOrName === 'string') return domainIdOrName;
+    else return this.domainIdToChainName(domainIdOrName);
+  }
+
+  /**
+   * Get an Ethers provider for a given chain name or id
+   */
+  tryGetProvider(chainNameOrId: ChainName | number): Provider | null {
+    const metadata = this.tryGetChainMetadata(chainNameOrId);
+    if (!metadata) return null;
+    const { name, id, publicRpcUrls } = metadata;
+
+    if (this.providers[name]) return this.providers[name];
+
+    if (TestChains.includes(name as CoreChainName)) {
+      this.providers[name] = new providers.JsonRpcProvider(
+        'http://localhost:8545',
+        31337,
+      );
+    } else if (publicRpcUrls.length) {
+      this.providers[name] = new providers.JsonRpcProvider(
+        publicRpcUrls[0].http,
```

Obviously don't have to this, but would love love for this to be a Fallback provider

---

Open to it but want to minimize semantic changes so punting on that

---

Different comment, since we don't validate that there actually are any URLs, we should throw an error here if it is undefined?

---

Unresolving this as it looks like we don't validator that there actually is a URL set? I.e. we should return `null` in that case?

---

Maybe even use `getRpcUrl`

---

You're looking at an outdated commit here. There is a check in the latest version